lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -1085,61 +1109,58 @@ CompilerType SwiftLanguageRuntimeImpl::GetChildCompilerTypeAtIndex(
         instance_type.GetTypeSystem().dyn_cast_or_null<TypeSystemSwift>();
     if (!instance_ts)
       return {};
-
     // LLDBTypeInfoProvider needs to be kept alive while supers gets accessed.
     llvm::SmallVector<SuperClassType, 2> supers;
-    LLDBTypeInfoProvider tip(*this, *instance_ts);
-    reflection_ctx->ForEachSuperClassType(
-        &tip, ts->GetDescriptorFinder(), pointer,
-        [&](SuperClassType sc) -> bool {
-          // If the typeref is invalid, we don't want to process it (for
-          // example, this could be an artifical ObjC class).
-          if (!sc.get_typeref())
-            return false;
+    auto superclass_finder = [&](SuperClassType sc) -> bool {
+      // If the typeref is invalid, we don't want to process it (for
+      // example, this could be an artifical ObjC class).
+      if (!sc.get_typeref())
+        return false;
 
-          if (!found_start) {
-            // The ValueObject always points to the same class instance,
-            // even when querying base classes. Drop base classes until we
-            // reach the requested type.
-            if (auto *tr = sc.get_typeref()) {
-              NodePointer base_class = tr->getDemangling(dem);
-              if (TypeSystemSwiftTypeRef::GetBaseName(base_class) != type_name)
-                return false;
-              found_start = true;
-            }
-          }
-          supers.push_back(sc);
-          return supers.size() >= 2;
-        });
+      if (!found_start) {
+        // The ValueObject always points to the same class instance,
+        // even when querying base classes. Drop base classes until we
+        // reach the requested type.
+        if (auto *tr = sc.get_typeref()) {
+          NodePointer base_class = tr->getDemangling(dem);
+          if (TypeSystemSwiftTypeRef::GetBaseName(base_class) != type_name)
+            return false;
+          found_start = true;
+        }
+      }
+      supers.push_back(sc);
+      return supers.size() >= 2;
+    };
 
-    if (supers.size() == 0) {
+    LLDBTypeInfoProvider tip(*this, *instance_ts);
+    // Try out the instance pointer based super class traversion first, as its
+    // usually faster.
+    reflection_ctx->ForEachSuperClassType(&tip, ts->GetDescriptorFinder(),
+                                          pointer, superclass_finder);
+
+    if (supers.empty())
+      // If the pointer based super class traversion failed (this may happen
+      // when metadata is not present in the binary, for example: embedded
+      // Swift), try the typeref based one next.
+      reflection_ctx->ForEachSuperClassType(&tip, ts->GetDescriptorFinder(), tr,
+                                            superclass_finder);
+
+    if (supers.empty()) {
```

Comment:
why twice the same condition in a row?

Reply:
Dynamic type resolution first tries the pointer-based `ForEachSuperClassType`, if that fails, it tries the typeref-based `ForEachSuperClassType`, if _that_ also fails then we're out of luck.

Reply:
I think my question was aiming at: does ForEachSuperClassType modify `supers`?
Otherwise why isn't this 
```
if (supers.empty()) {
   ForEachSuperClassType ...;
  LLDB_LOG
}
```

Reply:
Yes, it's being captured by reference by the `superclass_finder` lambda